main.go

```diff
@@ -30,80 +29,207 @@ func (fd fdReader) Read(p []byte) (n int, err error) {
-	}
+		err = serve(sessionInfo)
+		if err == nil {
+			log.Println("Exiting cleanly...")
+			os.RemoveAll(sessionInfo.path)
+		} else {
+			log.Println(err)
+			os.Exit(1)
+		}
+	case "new":
+		if parentSessionID != "" {
+			refuseNesting()
+			os.Exit(1)
+		}
+		if len(os.Args) != 3 {
+			fmt.Println("Usage: 3mux new <name>")
+			os.Exit(1)
+		}
+		sessionName := os.Args[2]
 
-	if *cmdHelp || *cmdHelpShort {
-		showHelp()
-		return
-	}
+		_, found, err := findSession(sessionName)
+		if err != nil {
+			fmt.Println("Error while querying sessions:", err)
+			os.Exit(1)
+		}
+		if found {
+			fmt.Println("Session with this name already exists:", sessionName)
+			os.Exit(1)
+		}
 
-	parentSessionID := os.Getenv("THREEMUX")
+		initializeSession(sessionName)
 
-	switch cmd {
-	case "help":
-		showHelp()
-	case "detach":
-		if parentSessionID == "" {
-			fmt.Println("Must be within session to detach")
-			return
+		daemonContext := &daemon.Context{
+			Args: []string{
+				os.Args[0],
+				"new-server-internal-only",
+				sessionName,
+			},
 		}
-		detach(parentSessionID)
-	case "_serve-id":
-		sessionID := os.Args[2]
-		serve(sessionID)
-	case "new":
+		child, err := daemonContext.Reborn()
+		if err != nil {
+			fmt.Println("Error occured while spawning session daemon:", err)
+			os.Exit(1)
+		}
+		if child == nil {
+			fmt.Println("Encountered 'unreachable' state with server-client mismatch. Args:", os.Args)
+			os.Exit(1)
+		}
+
+		os.Args[1] = "attach"
+		fallthrough
+	case "attach":
 		if parentSessionID != "" {
 			refuseNesting()
-			return
+			os.Exit(1)
+		}
+		if len(os.Args) != 3 {
+			fmt.Println("Usage: 3mux attach <name>")
+			os.Exit(1)
+		}
+		sessionName := os.Args[2]
+		sessionInfo, found, err := findSession(sessionName)
+		if err != nil {
+			fmt.Println("Error while querying sessions:", err)
+			os.Exit(1)
+		}
+		if !found {
+			fmt.Println("Failed to find session with name:", sessionName)
+			os.Exit(1)
+		}
+		err = attach(sessionInfo)
+		if err != nil {
+			fmt.Println(err.Error())
+			fmt.Println("See server-side logs at", path.Join(sessionInfo.path, "logs-server.txt"))
+			fmt.Printf("To manually kill this session, run `3mux kill %s`\n", sessionName)
+			os.Exit(1)
+		}
+	case "kill":
+		if len(os.Args) != 3 {
+			fmt.Println("Usage: 3mux kill <name>")
+			os.Exit(1)
 		}
 		sessionName := os.Args[2]
-		sessionID := launchServer(sessionName)
-		attach(sessionID)
+		sessionInfo, found, err := findSession(sessionName)
+		if err != nil {
+			fmt.Println("Error while querying sessions:", err)
+			os.Exit(1)
+		}
+		if !found {
+			fmt.Println("Failed to find session with name:", sessionName)
+			os.Exit(1)
+		}
+
+		_, err = net.Dial("unix", sessionInfo.killServerPath)
+		if err != nil {
+			fmt.Println("Killing the server failed.")
+			fmt.Println("To create a new session with this name:")
+			fmt.Println("1. Ensure there are no unwanted 3mux processes running")
+			fmt.Printf("2. Run `rm -rf %s`\n", sessionInfo.path)
```

Comment:
I'm not sure if this is necessary since there is already a `RemoveAll` on line 226